tests/jobs/test_scheduler_job.py

```diff
@@ -3869,13 +3870,112 @@ def dict_from_obj(obj):
+
+        ds_id = session.query(DatasetModel.id).filter_by(uri=dataset.uri).scalar()
+
+        event1 = DatasetEvent(
+            dataset_id=ds_id,
+            source_task_id="task",
+            source_dag_id=dr.dag_id,
+            source_run_id=dr.run_id,
+            source_map_index=-1,
+        )
+        session.add(event1)
+
+        # Create a second event, creation time is more recent, but data interval is older
+        dr = dag_maker.create_dagrun(
+            run_id="run2",
+            execution_date=(DEFAULT_DATE + timedelta(days=101)),
+            data_interval=(DEFAULT_DATE + timedelta(days=5), DEFAULT_DATE + timedelta(days=6)),
+        )
+
+        event2 = DatasetEvent(
+            dataset_id=ds_id,
+            source_task_id="task",
+            source_dag_id=dr.dag_id,
+            source_run_id=dr.run_id,
+            source_map_index=-1,
+        )
+        session.add(event2)
+
+        # Create DAG after dataset events.
+        with dag_maker(dag_id="datasets-consumer", schedule=[dataset]):
+            pass
+        dag = dag_maker.dag
 
-        assert dag3.get_last_dagrun().creating_job_id == scheduler_job.id
+        with dag_maker(dag_id="datasets-1-new", start_date=timezone.utcnow(), session=session):
+            BashOperator(task_id="task", bash_command="echo 1", outlets=[dataset])
+        dr = dag_maker.create_dagrun(
+            run_id="run3",
+            execution_date=(DEFAULT_DATE + timedelta(days=101)),
+            data_interval=(DEFAULT_DATE + timedelta(days=5), DEFAULT_DATE + timedelta(days=6)),
+        )
+
+        event3 = DatasetEvent(
+            dataset_id=ds_id,
+            source_task_id="task",
+            source_dag_id=dr.dag_id,
+            source_run_id=dr.run_id,
+            source_map_index=-1,
+            timestamp=timezone.utcnow(),
+        )
+        session.add(event3)
+
+        session = dag_maker.session
+        session.add(DatasetDagRunQueue(dataset_id=ds_id, target_dag_id=dag.dag_id))
+        session.flush()
+
+        scheduler_job = Job(executor=self.null_exec)
+        self.job_runner = SchedulerJobRunner(job=scheduler_job)
+
+        self.job_runner.processor_agent = mock.MagicMock()
+
+        with create_session() as session:
+            self.job_runner._create_dagruns_for_dags(session, session)
+
+        def dict_from_obj(obj):
+            """Get dict of column attrs from SqlAlchemy object."""
+            return {k.key: getattr(obj, k) for k in obj.__mapper__.column_attrs}
+
+        # dag should be triggered since it only depends on dataset1, it's been queued and dataset events landed after DAG was created.
+        created_run = session.query(DagRun).filter(DagRun.dag_id == dag.dag_id).one()
+        assert created_run.state == State.QUEUED
+
+        # we don't have __eq__ defined on DatasetEvent because... given the fact that in the future
+        # we may register events from other systems, dataset_id + timestamp might not be enough PK
```

Comment:
```suggestion
        # __eq__ isn't defined on DatasetEvent
```
IMO this is sufficient